core/trino-main/src/main/java/io/trino/sql/gen/columnar/DynamicPageFilter.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.gen.columnar;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.operator.project.SelectedPositions;
+import io.trino.spi.Page;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.DynamicFilter;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.spi.type.TypeManager;
+import io.trino.sql.PlannerContext;
+import io.trino.sql.ir.Expression;
+import io.trino.sql.ir.optimizer.IrExpressionOptimizer;
+import io.trino.sql.planner.DomainTranslator;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.relational.RowExpression;
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import jakarta.annotation.Nullable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.trino.sql.gen.columnar.FilterEvaluator.createColumnarFilterEvaluator;
+import static io.trino.sql.ir.optimizer.IrExpressionOptimizer.newOptimizer;
+import static io.trino.sql.relational.SqlToRowExpressionTranslator.translate;
+import static java.util.Objects.requireNonNull;
+
+public final class DynamicPageFilter
```

Comment:
For expressions there exists `io.trino.sql.gen.columnar.ColumnarFilterCompiler#filterCache`. Shouldn't there be same for DFs?

Reply:
DynamicPageFilter exists per scan. It is responsible for re-using same compiled class (modulo changes to DF) across splits for each scan.
Since we use the same underlying framework as static filter evaluation, dynamic filter RowExpressions will get cached in `ColumnarFilterCompiler#filterCache` through the calls to `createColumnarFilterEvaluator` in `createDynamicFilterEvaluator`. This will be re-usable across queries and in both static and dynamic filters.